libfreerdp/core/gateway/http.c

```diff
@@ -950,60 +1244,112 @@ HttpResponse* http_response_recv(rdpTls* tls, BOOL readContentLength)
 		{
-			int status;
-
-			if (!Stream_EnsureRemainingCapacity(response->data, bodyLength - response->BodyLength))
-				goto out_error;
+			char buffer[2048];
+			http_encoding_chunked_context ctx;
+			WLog_DBG(TAG, "Found a lazy web api developer that can't figure out content-length "
+			              "and uses chunked encoding instead");
+			ctx.state = ChunkStateLenghHeader;
+			ctx.nextOffset = 0;
+			ctx.headerFooterPos = 0;
+			int full_len = 0;
+			do
+			{
+				int status = http_chuncked_read(tls->bio, (BYTE*)buffer, sizeof(buffer), &ctx);
+				if (status <= 0)
+				{
+					if (!BIO_should_retry(tls->bio))
+					{
+						WLog_ERR(TAG, "Retries exceeded");
+						ERR_print_errors_cb(print_bio_error, NULL);
+						goto out_error;
+					}
+
+					USleep(100);
+				}
+				else
+				{
+					if (!Stream_EnsureRemainingCapacity(response->data, status))
+						goto out_error;
 
-			ERR_clear_error();
-			status = BIO_read(tls->bio, Stream_Pointer(response->data),
-			                  bodyLength - response->BodyLength);
+					memcpy(Stream_Pointer(response->data), buffer, status);
+					Stream_Seek(response->data, (size_t)status);
+					full_len += status;
+				}
 
-			if (status <= 0)
+			} while (ctx.state != ChunkStateEnd);
+			response->BodyLength = full_len;
+			if (response->BodyLength > 0)
+				response->BodyContent = &(Stream_Buffer(response->data))[payloadOffset];
+		}
+		else
+		{
+			while (response->BodyLength < bodyLength)
 			{
-				if (!BIO_should_retry(tls->bio))
-				{
-					WLog_ERR(TAG, "Retries exceeded");
-					ERR_print_errors_cb(print_bio_error, NULL);
+				int status;
+
+				if (!Stream_EnsureRemainingCapacity(response->data,
+				                                    bodyLength - response->BodyLength))
 					goto out_error;
-				}
 
-				USleep(100);
-				continue;
-			}
+				ERR_clear_error();
+				status = BIO_read(tls->bio, Stream_Pointer(response->data),
+				                  bodyLength - response->BodyLength);
 
-			Stream_Seek(response->data, (size_t)status);
-			response->BodyLength += (unsigned long)status;
+				if (status <= 0)
+				{
+					if (!BIO_should_retry(tls->bio))
+					{
+						WLog_ERR(TAG, "Retries exceeded");
+						ERR_print_errors_cb(print_bio_error, NULL);
+						goto out_error;
+					}
+
+					USleep(100);
+					continue;
+				}
 
-			if (response->BodyLength > RESPONSE_SIZE_LIMIT)
-			{
-				WLog_ERR(TAG, "Request body too large! (%" PRIdz " bytes) Aborting!",
-				         response->BodyLength);
-				goto out_error;
+				Stream_Seek(response->data, (size_t)status);
+				response->BodyLength += (unsigned long)status;
+
+				if (response->BodyLength > RESPONSE_SIZE_LIMIT)
+				{
+					WLog_ERR(TAG, "Request body too large! (%" PRIdz " bytes) Aborting!",
+					         response->BodyLength);
+					goto out_error;
+				}
 			}
-		}
 
-		if (response->BodyLength > 0)
-			response->BodyContent = &(Stream_Buffer(response->data))[payloadOffset];
+			if (response->BodyLength > 0)
```

Comment:
is the buffer checked for `payloadOffset` bytes available?

Reply:
this code is not really mine, but payload offset is taken from a Stream_GetPosition

It may make sense adding a WINPR_ASSERT

Reply:
@akarl10 ah, ok. did miss that part, then it is already `checked`.